File: 1202/dfs.py
from typing import *


class Solution:
    def smallestStringWithSwaps(self, s: str, pairs: List[List[int]]) -> str:
        n = len(s)
        neighbors = {i: [] for i in range(n)}
        for i1, i2 in pairs:
            neighbors[i1].append(i2)
            neighbors[i2].append(i1)

        ret = list(s)
        visited = [False for _ in range(n)]
        for index in range(n):
            if visited[index]:
                continue
            indices = []
            self.dfs(index, neighbors, visited, indices)
            chars = [s[index] for index in indices]
            indices.sort()
            chars.sort()
            for idx, char in zip(indices, chars):
                ret[idx] = char
        return ''.join(ret)
   
    # Get all indices of the same group
    def dfs(self, index: int, neighbors: Dict[int, List[int]], visited: List[bool], indices: List[int]) -> None:
        visited[index] = True
        indices.append(index)
        for neighbor in neighbors[index]:
            if visited[neighbor]:
                continue
            self.dfs(neighbor, neighbors, visited, indices)


if __name__ == '__main__':
    sol = Solution()

    s, pairs, expected = "dcab", [[0, 3], [1, 2]], "bacd"
    actual = sol.smallestStringWithSwaps(s, pairs)
    assert actual == expected

    s, pairs, expected = "dcab", [[0, 3], [1, 2], [0, 2]], "abcd"
    actual = sol.smallestStringWithSwaps(s, pairs)
    assert actual == expected

    s, pairs, expected = "cba", [[0, 1], [1, 2]], "abc"
    actual = sol.smallestStringWithSwaps(s, pairs)
    assert actual == expected

    s, pairs, expected = "dcab", [], "dcab"
    actual = sol.smallestStringWithSwaps(s, pairs)
    assert actual == expected

File: 1202/dfs.rs
use std::collections::HashMap;

struct Solution;

// TODO: TLE
impl Solution {
    pub fn smallest_string_with_swaps(s: String, pairs: Vec<Vec<i32>>) -> String {
        let mut neighbors = HashMap::new();
        for pair in pairs {
            let (i0, i1) = (pair[0] as usize, pair[1] as usize);
            neighbors.entry(i0).or_insert(Vec::new()).push(i1);
            neighbors.entry(i1).or_insert(Vec::new()).push(i0);
        }
        let mut ret: Vec<char> = vec!['a'; s.len()];
        let mut visited: Vec<bool> = vec![false; s.len()];
        for index in 0..s.len() {
            if visited[index] {
                continue;
            }
            // isolated node
            if !neighbors.contains_key(&index) {
                ret[index] = s.chars().nth(index).unwrap();
                continue;
            }
            let mut indices: Vec<usize> = Vec::new();
            Solution::dfs(index, &neighbors, &mut visited, &mut indices);
            let mut chars: Vec<char> = indices.iter().map(|i| s.chars().nth(*i).unwrap()).collect();
            indices.sort();
            chars.sort();
            for i in 0..indices.len() {
                ret[indices[i]] = chars[i];
            }
        }
        return ret.into_iter().collect();
    }

    // Get all indices of the same group
    pub fn dfs(index: usize, neighbors: &HashMap<usize, Vec<usize>>, visited: &mut Vec<bool>, indices: &mut Vec<usize>){
        visited[index] = true;
        indices.push(index);
        for neighbor in neighbors.get(&index).unwrap() {
            if visited[*neighbor] {
                continue;
            }
            Solution::dfs(*neighbor, neighbors, visited, indices);
        }
    }
}

fn main() {
    let (s, pairs, expected) = (String::from("dcab"), vec!{vec!{0, 3}, vec!{1, 2}}, String::from("bacd"));
    let actual = Solution::smallest_string_with_swaps(s, pairs);
    assert_eq!(actual, expected);

    let (s, pairs, expected) = (String::from("dcab"), vec!{vec!{0, 3}, vec!{1, 2}, vec!{0, 2}}, String::from("abcd"));
    let actual = Solution::smallest_string_with_swaps(s, pairs);
    assert_eq!(actual, expected);

    let (s, pairs, expected) = (String::from("cba"), vec!{vec!{0, 1}, vec!{1, 2}}, String::from("abc"));
    let actual = Solution::smallest_string_with_swaps(s, pairs);
    assert_eq!(actual, expected);

    let (s, pairs, expected) = (String::from("dcab"), Vec::new(), String::from("dcab"));
    let actual = Solution::smallest_string_with_swaps(s, pairs);
    assert_eq!(actual, expected);
}
